### PR TITLE
[Lock] Unlock CRU18, CRU24 or CRU24D6 Module

### PR DESCRIPTION
**Pull request by MyronFanQiu**

### CODE REVIEW

File: cstrml/locks/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+            Ok(())
+        }
+
+        /// Unlock the CRU18 or CRU24 one period
+        #[weight = 1000]
+        fn unlock_one_period(origin) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            // Ensure the start date is set and who has the CRU18 or CRU24
+            ensure!(Self::start_date().is_some(), Error::<T>::NotStarted);
```
Comment (zikunfan) [inline cstrml/locks/src/lib.rs:145]: Add `curr_bn >= start_date`

File: cstrml/locks/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            StartDate::put(date);
+
+            Self::deposit_event(RawEvent::SetStartDateSuccess(date));
+
+            Ok(())
+        }
+
+        /// Unlock the CRU18 or CRU24 one period
+        #[weight = 1000]
+        fn unlock_one_period(origin) -> DispatchResult {
```
Comment (zikunfan) [inline cstrml/locks/src/lib.rs:141]: Support unlock multiple periods

**Review by zikunfan — CHANGES_REQUESTED** [review]

File: cstrml/locks/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+                        last_unlock_at: 0,
+                        lock_type
+                    })
+                }
+            }
+
+        });
+
+        let total_amount = Self::locks(&who).unwrap().total;
+        T::Currency::set_lock(
```
Comment (zikunfan) [inline cstrml/locks/src/lib.rs:234]: Just set lock without issue CRU?

File: cstrml/locks/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            };
+
+            ensure!(curr_period > target_unlock_period, Error::<T>::TimeIsNotEnough);
+
+            // Count the total unlock period => Count the total unlock amount => Refresh the remaining locked amount
+            let free_periods = curr_period.saturating_sub(unlock_date).saturating_sub(lock.lock_type.delay) / T::UnlockPeriod::get();
+            let free_amount = Perbill::from_rational_approximation(free_periods, lock.lock_type.lock_period) * lock.total;
+            let locked_amount = lock.total - free_amount;
+            Self::update_lock(&who, locked_amount);
+            lock.last_unlock_at = curr_period;
```
Comment (zikunfan) [inline cstrml/locks/src/lib.rs:172]: Add comments with steps, and merge 172 and 174 as `Update Locks`

**Review by zikunfan — CHANGES_REQUESTED** [review]

File: cstrml/locks/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+            let target_unlock_period = if lock.last_unlock_at == 0 {
+                unlock_date + lock.lock_type.delay
+            } else {
+                lock.last_unlock_at
+            };
+
+            ensure!(curr_period > target_unlock_period, Error::<T>::TimeIsNotEnough);
+
+            // Count the total unlock period => Count the total unlock amount => Refresh the remaining locked amount
+            let free_periods = curr_period.saturating_sub(unlock_date).saturating_sub(lock.lock_type.delay) / T::UnlockPeriod::get();
```
Comment (zikunfan) [inline cstrml/locks/src/lib.rs:168]: `unlock_peroids` and `unlock_amount`

**Review by zikunfan — CHANGES_REQUESTED** [review]

File: cstrml/locks/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    }
+}
+
+impl<T: Config> Module<T> {
+    fn get_current_block_number() -> BlockNumber {
+        let current_block_number = <frame_system::Module<T>>::block_number();
+        TryInto::<u32>::try_into(current_block_number).ok().unwrap()
+    }
+
+    fn update_lock(who: &T::AccountId, locked_amount: BalanceOf<T>) {
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/locks/src/lib.rs]: changed it back

**Review by zikunfan — APPROVED** [review]
lgtm

**Pull request merged**